src/parser.rs

```diff
@@ -53,41 +55,41 @@ impl ReadLoopState {
 
     fn send_pong(&self) -> io::Result<()> {
         let w = &mut self.writer.lock().unwrap().writer;
-        w.write(b"PONG\r\n")?;
+        w.write_all(b"PONG\r\n")?;
         w.flush()?;
         Ok(())
     }
-}
-
-fn process_msg(state: &mut ReadLoopState, msg_args: MsgArgs) -> io::Result<()> {
-    let mut msg = Message {
-        subject: msg_args.subject,
-        reply: msg_args.reply,
-        data: Vec::with_capacity(msg_args.mlen as usize),
-        writer: None,
-    };
 
-    // Setup so we can send responses.
-    if let Some(_) = msg.reply {
-        msg.writer = Some(state.writer.clone());
-    }
+    fn process_msg(&mut self, msg_args: MsgArgs) -> io::Result<()> {
+        let mut msg = Message {
+            subject: msg_args.subject,
+            reply: msg_args.reply,
+            data: Vec::with_capacity(msg_args.mlen as usize),
+            writer: None,
+        };
+
+        // Setup so we can send responses.
+        if msg.reply.is_some() {
+            msg.writer = Some(self.writer.clone());
+        }
 
-    let reader = &mut state.reader;
-    // FIXME(dlc) - avoid copy if possible.
-    // FIXME(dlc) - Just read CRLF? Buffered so should be ok.
-    reader
-        .take(msg_args.mlen as u64)
-        .read_to_end(&mut msg.data)?;
+        let reader = &mut self.reader;
+        // FIXME(dlc) - avoid copy if possible.
+        // FIXME(dlc) - Just read CRLF? Buffered so should be ok.
+        reader
+            .take(u64::from(msg_args.mlen))
```

Comment:
Although the clippy lints that were added will protect a lot from it, `as` is generally good to avoid where possible. It's the source of a lot of accidental truncation etc... and it's a good practice for any sort of numerical work to use `from` and `try_from` to ensure you're never truncating.

Reply:
Why the extra indentation? How does behavior differ with u64::from?

Reply:
The indentation comes from the fact that I moved the function that was associated with the ReadLoopState to being a method defined on it, and placed in the impl block instead of being a function outside the block.

`try_from` differs from `from` by returning a `Result` if the conversion fails. So, if you try to do `u8::try_from(u64::max_value())` it will be an `Err` because it would truncate.

Reply:
Why don't you need the `?` at the end then?

Reply:
If using `try_from` instead of `try` you do need to handle errors (or blow up with an `expect` if it violates a core assumption that the value falls within the acceptable range). When using `u64::from` it doesn't return a `Result` but just a `u64` so there's no error to handle.

Reply:
ok so probably a naive question but why better than `as u64`?